src/plugins/plugin-koa.ts

```diff
@@ -13,134 +13,170 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-'use strict';
 
-const shimmer = require('shimmer');
-const urlParse = require('url').parse;
+// TODO(kjin): Remove this when @types/shimmer are published.
+// tslint:disable-next-line:no-reference
+/// <reference path="../types.d.ts" />
 
-function startSpanForRequest(api, req, res, next) {
+import {IncomingMessage, ServerResponse} from 'http';
+import * as shimmer from 'shimmer';
+import {parse as urlParse} from 'url';
+
+import {PluginTypes} from '..';
+
+import {koa_1, koa_2} from './types';
+
+type Koa1Module = typeof koa_1;
+type Koa2Module = typeof koa_2;
+// routePath is populated if the user uses the koa-route module.
+type KoaContext = (koa_1.Context|koa_2.Context)&{routePath?: string | RegExp};
+
+interface KoaModule<T> {
+  // TypeScript isn't expressive enough, but KoaModule#use should return `this`.
+  // tslint:disable-next-line:no-any
+  readonly prototype: {use: (m: T) => any};
```

Comment:
I'm not sure but would something like this work?

```typescript
readonly prototype: {use: (this: any, m: T) => this};
```

Reply:
The best we can do is

```ts
readonly prototype: {use: (this: any, m: T) => typeof this};
```

which loses the information that it returns `this`.